### PR TITLE
define sort and filter rules

### PR DESCRIPTION
**Pull request by Yahhi**
This will define sort and filter rules to be applied to requests to abstract product repository. Now it is ready to be used in bloc

### CODE REVIEW

File: lib/data/abstract/model/sort_rules.dart [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        break;
+      case SortType.Popularity:
+        if (sortOrder == SortOrder.FromHighestToLowest) {
+          return 'Popular first';
+        } else {
+          return 'Unpopular first';
+        }
+        break;
+      default:
+        return 'No order';
```
Comment (4seer) [inline lib/data/abstract/model/sort_rules.dart:50]: We need to have latest or most popular first if there is no order.

File: lib/data/abstract/model/sort_rules.dart [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    }
+  }
+}
+
+enum SortType { DateAdded, Reviewed, Price, Popularity }
+enum SortOrder { FromLowestToHighest, FromHighestToLowest }
```
Comment (4seer) [inline lib/data/abstract/model/sort_rules.dart:71]: How do  you plan to identify it in UI?

File: lib/data/fake_model/fake_product_repository.dart [hunk 3]
```diff
@@ -130,7 +124,33 @@ class FakeProductRepository extends ProductRepository {
...
 
+  @override
+  Future<FilterRules> getPossibleFilterOptions(int categoryId) async {
+    HashMap<ProductAttribute, List<String>> result = HashMap();
+    _productsInside.values.forEach((product) => result
+        .addAll({for (var attribute in product.attributes) attribute: []}));
+    return FilterRules(
+        categories: HashMap(),
+        selectedAttributes: result,
+        selectedPriceRange: PriceRange(10, 100));
```
Comment (4seer) [inline lib/data/fake_model/fake_product_repository.dart]: We need to make it according to the product overall price range.

File: lib/data/abstract/product_repository.dart [hunk 4]
```diff
@@ -16,13 +18,27 @@ abstract class ProductRepository {
...
+  ///The result can be filtered by
+  ///a) which category (with [categoryId] and its subcategories) they belong
+  ///b) [filterRules] which basically consist of attributes, subcategories and
+  ///price range, but can be extended to include more
+  ///c) only favorite products which were predefined by the user
+  ///The result can be sorted by [sortRules] with values coming from lowest to
+  ///highest or vice versa with main sort parameter specified.
+  Future<List<Product>> getProducts({
+    int pageIndex = 0,
+    int pageSize = AppConsts.PAGE_SIZE,
```
Comment (4seer) [inline lib/data/abstract/product_repository.dart]: Please, use camel case for constant names https://dart.dev/guides/language/effective-dart/style

File: lib/data/fake_model/fake_product_repository.dart (hunk 3 above)
Comment (4seer) [reply lib/data/fake_model/fake_product_repository.dart]: https://pub.dev/packages/font_awesome_flutter

https://fontawesome.com/icons?d=gallery&q=sort

File: lib/data/local/local_product_repository.dart [hunk 5]
```diff
@@ -28,16 +24,27 @@ class ProductLocalRepository implements ProductRepository {
   }
 
   @override
-  Future<List<Product>> getProductsInCategory(int categoryId,
+  Future<List<Product>> getSimilarProducts(int categoryI,
       {int pageIndex = 0, int pageSize = AppConsts.PAGE_SIZE}) {
```
Comment (4seer) [inline lib/data/local/local_product_repository.dart]: Change to camelCase for constants, please.

File: lib/presentation/features/products/bloc_list_data.dart [hunk 6]
```diff
@@ -0,0 +1,15 @@
+import 'package:openflutterecommerce/data/abstract/model/category.dart';
+import 'package:openflutterecommerce/data/abstract/model/hashtag.dart';
+import 'package:openflutterecommerce/data/abstract/model/product.dart';
+
+class ProductListData {
```
Comment (4seer) [inline lib/presentation/features/products/bloc_list_data.dart:5]: May be we should include here filter options too: price range, available colors, sizes and brands?

**Review by 4seer — CHANGES_REQUESTED** [review]
Please, make constant names camelCased and provide answers to the added comments.

File: lib/data/abstract/model/sort_rules.dart (hunk 1 above)
Comment (Yahhi) [reply lib/data/abstract/model/sort_rules.dart:50]: It is created with a default order and it is not null, so default here is just to be compatible with lint rules

File: lib/data/abstract/model/sort_rules.dart (hunk 2 above)
Comment (Yahhi) [reply lib/data/abstract/model/sort_rules.dart:71]: I am not sure yet. I think in the list here there should be an icon on right side to show the order
image

File: lib/data/fake_model/fake_product_repository.dart (hunk 3 above)
Comment (Yahhi) [reply lib/data/fake_model/fake_product_repository.dart]: I think for fake repository it is OK. For real database we should get real data

File: lib/data/abstract/model/sort_rules.dart (hunk 2 above)
Comment (4seer) [reply lib/data/abstract/model/sort_rules.dart:71]: Added as a separate issue here https://github.com/4seer/openflutterecommerceapp/issues/109

File: lib/data/local/local_product_repository.dart [hunk 7]
```diff
@@ -28,16 +24,27 @@ class ProductLocalRepository implements ProductRepository {
   }
 
   @override
-  Future<List<Product>> getProductsInCategory(int categoryId,
-      {int pageIndex = 0, int pageSize = AppConsts.PAGE_SIZE}) {
-    // TODO: implement getProductsInCategory
+  Future<List<Product>> getSimilarProducts(int categoryI,
```
Comment (4seer) [inline lib/data/local/local_product_repository.dart]: categoryId ?

File: lib/data/woo_commerce/product_remote_repository.dart [hunk 8]
```diff
@@ -28,16 +24,27 @@ class RemoteProductRepository extends ProductRepository {
   }
 
   @override
-  Future<List<Product>> getProductsInCategory(int categoryId,
-      {int pageIndex = 0, int pageSize = AppConsts.PAGE_SIZE}) {
-    // TODO: implement getProductsInCategory
+  Future<List<Product>> getSimilarProducts(int categoryI,
```
Comment (4seer) [inline lib/data/woo_commerce/product_remote_repository.dart]: categoryId ?

File: lib/presentation/features/favorites/favorite_bloc.dart [hunk 9]
```diff
@@ -0,0 +1,52 @@
+// Favourites Screen Bloc
```
Comment (4seer) [inline lib/presentation/features/favorites/favorite_bloc.dart]: Lets rename the file back to favorite**s**_bloc.dart

File: lib/presentation/features/favorites/views/favourites_tile_view.dart [hunk 10]
```diff
@@ -5,118 +5,51 @@
...
 
-import '../../wrapper.dart';
-import '../favorites_bloc.dart';
+import '../favorite_bloc.dart';
 import '../favorites_event.dart';
 import '../favorites_state.dart';
 
 class FavouritesTileView extends StatefulWidget {
-  final Function({@required ViewChangeType changeType, int index}) changeView;
+  final Function changeView;
```
Comment (4seer) [inline lib/presentation/features/favorites/views/favourites_tile_view.dart]: Why did you decide to remove the parameters?

File: lib/presentation/features/products/products_screen.dart [hunk 11]
```diff
@@ -44,34 +44,46 @@ class _ProductsScreenState extends State<ProductsScreen> {
...
+                    },
+                    onViewChanged: () {
+                      BlocProvider.of<ProductsBloc>(context)
+                          .add(ProductsChangeViewEvent());
+                    },
+                  ),
+                  state is ProductsListViewState
+                      ? ProductsListView()
+                      : ProductsTileView(),
+                ],
```
Comment (4seer) [inline lib/presentation/features/products/products_screen.dart:82]: It is not the best idea to render the screen depending on the state of the app.

File: lib/presentation/features/favorites/views/favourites_tile_view.dart (hunk 10 above)
Comment (Yahhi) [reply lib/presentation/features/favorites/views/favourites_tile_view.dart]: They are not required here. The view is changed by state

File: lib/presentation/features/products/bloc_list_data.dart (hunk 6 above)
Comment (Yahhi) [reply lib/presentation/features/products/bloc_list_data.dart:5]: I mean here that this data is shown on screen somehow. Filters and sort rules are just rules and are not shown in this exact screen. But in fact, it is not necessary to wrap it inside this.

**Review by 4seer — APPROVED** [review]

**Pull request merged**